Crash on startup - signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xf13031

```
/usr/bin/ipfs-cluster-service daemon 
2022-11-13T17:31:16.106+0400	INFO	service	ipfs-cluster-service/daemon.go:49	Initializing. For verbose output run with "-l debug". Please wait...
2022-11-13T17:31:16.126+0400	INFO	badger	badger@v1.6.2/logger.go:46	All 0 tables opened in 0s

2022-11-13T17:31:16.126+0400	INFO	badger	badger@v1.6.2/logger.go:46	Replaying file id: 0 at offset: 0

2022-11-13T17:31:16.126+0400	INFO	badger	badger@v1.6.2/logger.go:46	Replay took: 5.866µs

2022-11-13T17:31:16.127+0400	INFO	service	ipfs-cluster-service/daemon.go:267	Datastore backend: badger
2022-11-13T17:31:16.161+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:137	IPFS Cluster v1.0.4 listening on:
        /ip4/10.0.0.10/tcp/9096/p2p/12D3KooWGtzjXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
        /ip4/127.0.0.1/tcp/9096/p2p/12D3KooWGtzjjXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX


2022-11-13T17:31:16.831+0400	INFO	pinsvcapi	common/api.go:460	PINSVCAPI (HTTP): /ip4/127.0.0.1/tcp/9097
2022-11-13T17:31:16.831+0400	INFO	restapi	common/api.go:460	RESTAPI (HTTP): /ip4/127.0.0.1/tcp/9094
2022-11-13T17:31:16.831+0400	INFO	ipfsproxy	ipfsproxy/ipfsproxy.go:331	IPFS Proxy: /ip4/127.0.0.1/tcp/9095 -> /ip4/10.0.0.10/tcp/5001
2022-11-13T17:31:16.833+0400	INFO	crdt	go-ds-crdt@v0.3.9/set.go:122	Tombstones have bloomed: 0 tombs. Took: 170.972µs
2022-11-13T17:31:16.833+0400	INFO	crdt	go-ds-crdt@v0.3.9/crdt.go:296	crdt Datastore created. Number of heads: 0. Current max-height: 0. Dirty: false
2022-11-13T17:31:16.833+0400	INFO	crdt	crdt/consensus.go:319	'trust all' mode enabled. Any peer in the cluster can modify the pinset.
2022-11-13T17:31:16.834+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:735	Cluster Peers (without including ourselves):
2022-11-13T17:31:16.834+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:742	    - 12D3KooWNBCB6jXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
2022-11-13T17:31:16.834+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:747	Waiting for IPFS to be ready...
2022-11-13T17:31:16.834+0400	INFO	crdt	go-ds-crdt@v0.3.9/crdt.go:505	store is marked clean. No need to repair
2022-11-13T17:31:16.836+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:756	IPFS is ready. Peer ID: QmawbFWECFk6jXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
2022-11-13T17:31:16.836+0400	INFO	cluster	ipfs-cluster@v1.0.4/cluster.go:764	** IPFS Cluster is READY **
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xf13031]

goroutine 754 [running]:
github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).findNodes(0xc000a2ea90, {0x1aaeb70?, 0xc004802e80})
	github.com/libp2p/go-libp2p@v0.22.0/p2p/host/autorelay/relay_finder.go:209 +0x51
github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background.func2()
	github.com/libp2p/go-libp2p@v0.22.0/p2p/host/autorelay/relay_finder.go:112 +0x66
created by github.com/libp2p/go-libp2p/p2p/host/autorelay.(*relayFinder).background
	github.com/libp2p/go-libp2p@v0.22.0/p2p/host/autorelay/relay_finder.go:110 +0x192
  home ~  /usr/bin/ipfs-cluster-service --version
ipfs-cluster-service version 1.0.4
```


**Additional information:**
 - OS: Linux/Ubuntu
 - IPFS Cluster version: 1.0.4
 - Installation method: binary from github
**Describe the bug:**

Constantly crashing on startup. 100% reproducible. Am behind a NAT gateway, and port is getting punched in my router correctly.

--- a/clusterhost.go
+++ b/clusterhost.go
@@ -3,6 +3,7 @@ package ipfscluster
 import (
 	"context"
 	"encoding/hex"
+	"time"
 
 	config "github.com/ipfs-cluster/ipfs-cluster/config"
 	ds "github.com/ipfs/go-datastore"
@@ -16,8 +17,10 @@ import (
 	crypto "github.com/libp2p/go-libp2p/core/crypto"
 	host "github.com/libp2p/go-libp2p/core/host"
 	network "github.com/libp2p/go-libp2p/core/network"
+	peer "github.com/libp2p/go-libp2p/core/peer"
 	corepnet "github.com/libp2p/go-libp2p/core/pnet"
 	routing "github.com/libp2p/go-libp2p/core/routing"
+	"github.com/libp2p/go-libp2p/p2p/host/autorelay"
 	connmgr "github.com/libp2p/go-libp2p/p2p/net/connmgr"
 	identify "github.com/libp2p/go-libp2p/p2p/protocol/identify"
 	noise "github.com/libp2p/go-libp2p/p2p/security/noise"
@@ -65,7 +68,23 @@ func NewClusterHost(
 		return nil, nil, nil, err
 	}
 
+	var h host.Host
 	var idht *dual.DHT
+	// a channel to wait until these variables have been set
+	// (or left unset on errors). Mostly to avoid reading while writing.
+	hostAndDHTReady := make(chan struct{})
+	defer close(hostAndDHTReady)
+
+	hostGetter := func() host.Host {
+		<-hostAndDHTReady // closed when we finish NewClusterHost
+		return h
+	}
+
+	dhtGetter := func() *dual.DHT {
+		<-hostAndDHTReady // closed when we finish NewClusterHost
+		return idht
+	}
+
 	opts := []libp2p.Option{
 		libp2p.ListenAddrs(cfg.ListenAddr...),
 		libp2p.NATPortMap(),
@@ -76,9 +95,7 @@ func NewClusterHost(
 		}),
 		libp2p.EnableNATService(),
 		libp2p.EnableRelay(),
-		// breaks without passing a way to discover relays.
-		// https://github.com/libp2p/go-libp2p/issues/1852
-		// libp2p.EnableAutoRelay(),
+		libp2p.EnableAutoRelay(autorelay.WithPeerSource(newPeerSource(hostGetter, dhtGetter), time.Minute)),
 		libp2p.EnableHolePunching(),
 	}
 
@@ -86,7 +103,7 @@ func NewClusterHost(
 		opts = append(opts, libp2p.EnableRelayService())
 	}
 
-	h, err := newHost(
+	h, err = newHost(
 		ctx,
 		cfg.Secret,
 		ident.PrivateKey,
@@ -163,6 +180,63 @@ func newPubSub(ctx context.Context, h host.Host) (*pubsub.PubSub, error) {
 		pubsub.WithMessageSigning(true),
 		pubsub.WithStrictSignatureVerification(true),
 	)
+}
+
+type peerSourceF func(ctx context.Context, numPeers int) <-chan peer.AddrInfo
+
+// Inspired in Kubo's
+// https://github.com/ipfs/go-ipfs/blob/9327ee64ce96ca6da29bb2a099e0e0930b0d9e09/core/node/libp2p/relay.go#L79-L103
+// and https://github.com/ipfs/go-ipfs/blob/9327ee64ce96ca6da29bb2a099e0e0930b0d9e09/core/node/libp2p/routing.go#L242-L317
+// but simplified and adapted:
+//   - Everytime we need peers for relays we do a DHT lookup.
+//   - We return the peers from that lookup.
+//   - No need to do it async, since we have to wait for the full lookup to
+//     return anyways. We put them on a buffered channel and be done.
+func newPeerSource(hostGetter func() host.Host, dhtGetter func() *dual.DHT) peerSourceF {
+	return func(ctx context.Context, numPeers int) <-chan peer.AddrInfo {
+		// make a channel to return, and put items from numPeers on
+		// that channel up to numPeers. Then close it.
+		r := make(chan peer.AddrInfo, numPeers)
+		defer close(r)
+
+		// Because the Host, DHT are initialized after relay, we need to
+		// obtain them indirectly this way.
+		h := hostGetter()
+		if h == nil { // context canceled etc.
+			return r
+		}
+		idht := dhtGetter()
+		if idht == nil { // context canceled etc.
+			return r
+		}
+
+		// length of closest peers is K.
+		closestPeers, err := idht.WAN.GetClosestPeers(ctx, h.ID().String())
+		if err != nil { // Bail out. Usually a "no peers found".
+			return r
+		}
+
+		logger.Debug("peerSource: %d closestPeers for %d requested", len(closestPeers), numPeers)
+
+		for _, p := range closestPeers {
+			addrs := h.Peerstore().Addrs(p)
+			if len(addrs) == 0 {
+				continue
+			}
+			dhtPeer := peer.AddrInfo{ID: p, Addrs: addrs}
+			// Attempt to put peers on r if we have space,
+			// otherwise return (we reached numPeers)
+			select {
+			case r <- dhtPeer:
+			case <-ctx.Done():
+				return r
+			default:
+				return r
+			}
+		}
+		// We are here if numPeers > closestPeers
+		return r
+	}
 }
 
 // EncodeProtectorKey converts a byte slice to its hex string representation.

--- a/clusterhost.go
+++ b/clusterhost.go
@@ -9,15 +9,15 @@ import (
 	namespace "github.com/ipfs/go-datastore/namespace"
 	ipns "github.com/ipfs/go-ipns"
 	libp2p "github.com/libp2p/go-libp2p"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	dual "github.com/libp2p/go-libp2p-kad-dht/dual"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	record "github.com/libp2p/go-libp2p-record"
 	crypto "github.com/libp2p/go-libp2p/core/crypto"
 	host "github.com/libp2p/go-libp2p/core/host"
 	network "github.com/libp2p/go-libp2p/core/network"
 	corepnet "github.com/libp2p/go-libp2p/core/pnet"
 	routing "github.com/libp2p/go-libp2p/core/routing"
-	dht "github.com/libp2p/go-libp2p-kad-dht"
-	dual "github.com/libp2p/go-libp2p-kad-dht/dual"
-	pubsub "github.com/libp2p/go-libp2p-pubsub"
-	record "github.com/libp2p/go-libp2p-record"
 	connmgr "github.com/libp2p/go-libp2p/p2p/net/connmgr"
 	identify "github.com/libp2p/go-libp2p/p2p/protocol/identify"
 	noise "github.com/libp2p/go-libp2p/p2p/security/noise"
@@ -76,7 +76,9 @@ func NewClusterHost(
 		}),
 		libp2p.EnableNATService(),
 		libp2p.EnableRelay(),
-		libp2p.EnableAutoRelay(),
+		// breaks without passing a way to discover relays.
+		// https://github.com/libp2p/go-libp2p/issues/1852
+		// libp2p.EnableAutoRelay(),
 		libp2p.EnableHolePunching(),
 	}
 

